repo2docker/contentproviders/zenodo.py

```diff
@@ -28,39 +30,78 @@ def _urlopen(self, req, headers=None):
 
         return urlopen(req)
 
-    def detect(self, doi, ref=None, extra_args=None):
-        """Trigger this provider for things that resolve to a Zenodo record"""
-        # To support Zenodo instances not hosted at zenodo.org we need to
-        # start maintaining a list of known DOI prefixes and their hostname.
-        # We should also change to returning a complete `record_url` that
-        # fetch() can use instead of constructing a URL there
-        doi = doi.lower()
-        # 10.5281 is the Zenodo DOI prefix
-        if doi.startswith("10.5281/"):
+    def _doi2url(self, doi):
+        # Transform a DOI to a URL
+        # If not a doi, assume we have a URL and return
+        if is_doi(doi):
+            doi = normalize_doi(doi)
+
             resp = self._urlopen("https://doi.org/{}".format(doi))
-            self.record_id = resp.url.rsplit("/", maxsplit=1)[1]
-            return {"record": self.record_id}
-
-        elif doi.startswith("https://doi.org/10.5281/") or doi.startswith(
-            "http://doi.org/10.5281/"
-        ):
-            resp = self._urlopen(doi)
-            self.record_id = resp.url.rsplit("/", maxsplit=1)[1]
-            return {"record": self.record_id}
-
-        elif doi.startswith("https://zenodo.org/record/") or doi.startswith(
-            "http://zenodo.org/record/"
-        ):
-            self.record_id = doi.rsplit("/", maxsplit=1)[1]
-            return {"record": self.record_id}
+            return resp.url
+        else:
+            return doi
+
+    def _getfromdict(self, datadict, dotpath):
+        # Use a dotpath (string separated by periods)
+        # to access vaules in a dictionary
+        # data.files.0 returns value at dataDict[data][files][0]
+        split = dotpath.split(".")
+        # We check if we have any digits and convert these to
+        # ints for list access
+        mapList = []
+        for s in split:
+            if s.isdigit():
+                mapList.append(int(s))
+            else:
+                mapList.append(s)
+        values = copy.deepcopy(datadict)
+        for k in mapList:
+            values = values[k]
+        return values
+
+    def detect(self, doi, ref=None, extra_args=None):
+        """Trigger this provider for things that resolve to a Zenodo/Invenio record"""
+        # We need the hostname (url where records are), api url (for metadata),
+        # filepath (path to files in metadata), filename (path to filename in
+        # metadata), type (path to type in metadata)
+
```

Comment:
I'd remove this to make it clear that the comment belongs to the `hosts` variable that follows.